oeps/oep-0036-microservice-routing-layer.rst

```diff
@@ -0,0 +1,99 @@
+===================================
+OEP-36: Microservice Routing Layer
+===================================
+
+.. list-table::
+
+  * - OEP
+    - :doc:`oep-0036-microservice-routing-layer`
+  * - Title
+    - Microservice Routing Layer
+  * - Last-Modified
+    - 2019-04-25
+  * - Author
+    - Fred Smith <derf@edx.org>
+  * - Arbiter
+    - 
+  * - Status
+    - Provisional
+  * - Type
+    - Architecture
+  * - Created
+    - 2019-04-25
+
+Abstract
+========
+
+Proposes a system with which users can interact with microservice applications
+using a layer 7 routing framework. 
+
+Motivation
+==========
+
+Microservices are currently implemented using layer 4 (DNS) based routing. Each
+microservice occupies a unique hostname, with dedicated ELB or CDN 
+infrastructure to support routing user traffic to the service. This pattern
+is simple to monitor, deploy, and manage, but slow and expensive to build,
+and regularly causes problems with cross service user information sharing. In
+Addition, modifying or splitting a monolithic service into microservices resuls
+in confusing URL changes and costly migrations.
+
+A layer 7 routing layer would allow us to separate the user facing URL or API
+endpoint from the backend implementation. It would allow microservices to exist
+that only manage a small number of endpoints, without requring developers to 
+build out load balancing or CDN infrastructure. 
+
+We are currently using a layer 7 routing framework for prospectus, allowing
+us to replace drupal generated pages with gatsby generated pages quickly and
+seamlessly.
+
+This change will allow developers, openedx operators and resellers to deploy
+new openedx sites quicker and easier than they currently can.  The router
+will provide a map for people to understand which services they need to deploy.
+
+
+Specification
+=============
+
+A layer 7 routing application will be chosen that meets the following
+requirements:
+
+* High performance connection management capabilities, including support
+  for slow clients or backends, and wire-speed throughput.
+* SSL termination on the router.
+* Flexible configuration backends including container service discovery,
+  flat files, and filesystem routing.
+* Path based routing, including path subsets and regular expression matching
+
+This layer 7 routing application will be deployed on a common hostname.
```

Comment:
Does this imply that we won't be able to use custom subdomains with this routing layer?

Reply:
I think a goal is to reduce hostname sprawl.  http/2 and jwt work better with a single hostname (I think) so unless there's a good reason we would discourage services moving to a custom subdomain.

Technically, there isn't a big problem with custom subdomains for, say, multi-tenant installations.

Reply:
I agree with the discourage to move MFE to new domains. Does this apply to more than just MFE, for example other IDAs?

Also, multi-tenant installations very often have a completely different domain. This sometimes is problematic for SSL termination, but most applications that deal with SSL termination support SNI natively.